### PR TITLE
[JUJU-1653] Add support for removing secrets

### PR DESCRIPTION
**Pull request by wallyworld**
Add support for removing secrets. A new `secret-remove` hook command is added.
We also add a new "scope" attribute to the secret metadata - when the scoped entity is removed, so too are any associated secrets. Right now, the scope is not able to be set - it is hard wired to the application which owns the secret.

Secret deletion is done using bulk collection removes, to clean up all the dependent records like permissions etc. To allow for this, extra checks are added to prevent a secret being created when the owner or scoped entity is not alive. And the top level parent metadata record is deleted first, plus the secret access methods have an extra check added and will return not found if the secret is gone.

Attempts to access a non existent secret result in permission denied.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- [X] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

```
juju deploy juju-qa-dummy-source
juju exec --unit dummy-source/0 "secret-add data=foo"
secret:cbu6ssdr22hi5hqcv3pg
juju exec --unit dummy-source/0 "secret-get secret:cbu6ssdr22hi5hqcv3pg"
data: foo
juju exec --unit dummy-source/0 "secret-remove secret:cbu6ssdr22hi5hqcv3pg"
juju exec --unit dummy-source/0 "secret-get secret:cbu6ssdr22hi5hqcv3pg"
ERROR permission denied
```
check app deletion also deletes secret
```
juju exec --unit dummy-source/0 "secret-add data=foo"
secret:cbu702tr22hi5hqcv3q0
juju secrets
URI                          Revision  Rotate  Backend  Age
secret:cbu702tr22hi5hqcv3q0         1  never   juju     3 seconds ago  

juju remove-application dummy-source
removing application dummy-source
juju secrets
URI  Revision  Rotate  Backend  Age

```

### CODE REVIEW

File: secrets/interface.go [hunk 1]
```diff
@@ -33,6 +34,10 @@ func (p *CreateParams) Validate() error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+	_, err = names.ParseTag(p.Scope)
```
Comment (hpidcock) [inline secrets/interface.go:37]: no scope type validation?

File: state/secrets.go [hunk 2]
```diff
@@ -321,8 +385,79 @@ func (s *secretsStore) toSecretMetadata(doc *secretMetadataDoc) (*secrets.Secret
...
 }
 
+// DeleteSecret deletes the specified secret.
+func (s *secretsStore) DeleteSecret(uri *secrets.URI) error {
+	// We will bulk delete the various artefacts, starting with the secret itself.
+	// Deleting the parent secret metadata first  will ensure that any consumers of
+	// the secret get notified and subsequent attempts to access any secret
+	// attributes (revision etc) return not found.
+	// It is not practical to do this record by record in a mgo txn operation.
+	secretMetadataCollection, closer := s.st.db().GetRawCollection(secretMetadataC)
```
Comment (hpidcock) [inline state/secrets.go:395]: why can't this be done in a txn?

File: worker/uniter/runner/jujuc/secret-remove.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+type secretRemoveCommand struct {
+	cmd.CommandBase
+	ctx Context
+
+	uri string
+}
+
+// NewSecretRemoveCommand returns a command to create a secret.
```
Comment (hpidcock) [inline worker/uniter/runner/jujuc/secret-remove.go:21]: -create
+remove

**Review by hpidcock — APPROVED** [review]

File: state/secrets.go (hunk 2 above)
Comment (wallyworld) [reply state/secrets.go:409]: added start/commit txn

**Pull request merged**